### PR TITLE
Update "lambda package" to use non-root user when launching Docker on Linux/MacOS

### PR DESCRIPTION
**Pull request by jasonterando**
This is related to [Issue #256](https://github.com/aws/aws-extensions-for-dotnet-cli/issues/256)

Description of Change:  Update `dotnet lambda package` so that it launches Docker as the logged in user when being run in Linux or a MacOS.  This will allow `dotnet lambda deploy-function` to successfully create the .zip package on Linux.  This change also has the benfit of not leaving directories around that users have to `chown` or login in as root to clean up.  This will allow non-root developers to successfully use `dotnet lambda` to package and deploy.

Recap of Issue:  When `dotnet lambda package` today builds a project for linux-x64, it launches a Docker container.  Within the container, commands are executed which create /obj and  /bin directories using the Docker container's effective user, which is "root".  When deploying the package, `dotnet lambda deploy-function` attempts to build a .zip within these created directories, and fails, unless the active user is logged in as "root", which is not a good thing to do when you are developing.

Implementation Details:

1. Add class **PosixUserHelper** to assist with running `id` to get the user's User ID and Group ID (unit test **PosixUserHelperTest.cs**)
2. Update **DockerCLIWrapper** to call PosixUserHelper to determine if running under Unix/MacOS and, if so, add "-u" parameter to specify user and group IDs.  Also, change the default locations of .NET and .NuGet to /tmp so that non-root users can install
3. Update **DockerCLIWrapper** to build a list of arguments using List<string> instead of individual variables that are concatenated together.  This will make it easier in the future to add new arguments

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false,
+                CreateNoWindow = true,
+            };
+            proc.EnableRaisingEvents = true;
+            proc.OutputDataReceived += (_, e) =>
+            {
+                if (string.IsNullOrEmpty(e.Data)) return;
+                var results = Regex.Match(e.Data, @"uid=(\d+).*gid=(\d+)");
```
Comment (Beau-Gosse-dev) [inline src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs:48]: Would it be cleaner and easier to maintain to call `id -u` and `id -g` to get the `uid` and `gid` directly without regex?

File: src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+
+namespace Amazon.Common.DotNetCli.Tools
+{
+    public struct PosixUserInfo
+    {
+        public uint UserID;
+        public uint GroupID;
```
[the remark this answers is not in the document]
Comment (Beau-Gosse-dev) [reply src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs:10]: You could also use nullable ints internally instead of checking `uint.MaxValue`. Although extremely unlikely, I guess technically a uid might be the max value.

File: src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            }
+
+            if (userID == uint.MaxValue)
+            {
+                throw new Exception("Unable to get effective user from \"id\"");
+            }
+
+            if (groupID == uint.MaxValue)
+            {
+                throw new Exception("Unable to get effective group from \"id\"");
```
Comment (Beau-Gosse-dev) [inline src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs:82]: @normj and team may want you to add to the [ToolExceptions ](https://github.com/aws/aws-extensions-for-dotnet-cli/blob/master/src/Amazon.Common.DotNetCli.Tools/ToolsException.cs#L10) and use those, although I'm not sure if we want to add all of these as different ones. Maybe we can log the details instead and have 1 new exception type related to GetEffectiveUser.

File: src/Amazon.Common.DotNetCli.Tools/DefaultConfigFile.cs [hunk 4]
```diff
@@ -133,6 +134,25 @@ public string GetValueAsString(CommandOption option)
             return null;
         }
 
+        /// <summary>
+        /// Gets the default if it exists as a dictionary.
+        /// </summary>
+        /// <param name="option"></param>
+        /// <returns></returns>
+        public IDictionary<string, string> GetValueAsDictionary(CommandOption option)
```
Comment (Beau-Gosse-dev) [inline src/Amazon.Common.DotNetCli.Tools/DefaultConfigFile.cs:142]: This could be a separate PR to the PosixUserInfo change, right?

**Review by Beau-Gosse-dev — CHANGES_REQUESTED** [review]
Thanks for the PR! I tested this with `dotnet lambda package` on Windows against an AOT project and it worked fine.

**Comment by jasonterando** [comment]
Hi, thanks for the feedback, and thanks for confirming it's good on Windows.  If I'm reading correctly, seems like I need to address the following:
1. Add comment to assignment of .NET CLI and NuGet home environment variables:  No problem, will do
2. Change uint to int, maybe make nullable:  Happy to set to int, but did not see Nullable set up in Amazon.Lambda.Tools.csproj, so I think I would have to keep track of whether they are set using a boolean.  Come to think of it, this is probably fine, I can set the defaults for uid and gid to zero (root) and display a warning for Linux or MacOS if I can't use `id` to get the info
3. If there's a preference to do two separate calls to get user and group ID versus a single parsed via regex, that's cool.  Whatever gets the PR approved :)
4. GetValueAsDictionary should have been a separate PR, I must have included this by mistake, will address.

Assuming I do #1, #2 and #4 -- and we decide on #3, is there anything else you need from me to proceed?  Thanks!

**Comment by Beau-Gosse-dev** [comment]
Thanks for the quick follow up @jasonterando! The only thing I would add is that there are still some changes for using dictionary in the CLI args. I didn't look at those changes too closely, but seems like all of those could go in another PR. I'll need one of the repo contributors to review this too, since I don't have write access to this repository.

**Comment by jasonterando** [comment]
Okee doke - I think I _finally_ have the dictionary/parameters thing completely cleared out (I learned my lesson, I will remember next time to not do a PR from local master branch, sorry about the hassle).  Let me know if there is anything else before you feed this up the food chain.

**Comment by Beau-Gosse-dev** [comment]
@normj Is there some from this repo who can review this? I also reproduced the problem on Ubuntu and verified that this fixes it.

File: src/Amazon.Common.DotNetCli.Tools/DockerCLIWrapper.cs [hunk 5]
```diff
@@ -163,18 +164,52 @@ public int Push(string targetTagName)
...
 
-            var imageArg = $"-i {imageId}";
+                "-i"
+            };
+            
+            // For Linux or MacOS, running a .NET image as non-root requires some special
+            // handling.  We have to pass the user's UID and GID, but we also have to specify
+            // the location of where .NET and Nuget put their files, because by default, they
+            // go to the /.dotnet and /.local, which a non-root user can't write to in a container
+            if (PosixUserHelper.IsRunningInPosix)
```
Comment (normj) [inline src/Amazon.Common.DotNetCli.Tools/DockerCLIWrapper.cs:189]: If the user id is `0` then I think we should skip this part since we are already running at root.

**Review by normj — CHANGES_REQUESTED** [review]

**Comment by normj** [comment]
> @normj Is there some from this repo who can review this? I also reproduced the problem on Ubuntu and verified that this fixes it.

Back from winter holidays and added comments.

File: src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        public static readonly bool IsRunningInPosix = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ||
+                                                       RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
+
+        /// <summary>
+        /// Return the effective user's UID and GID under Linux and Mac by calling the "id" command.
+        /// This will fault if running on Windows (by design), check IsRunningInPosix before calling this method.
+        /// </summary>
+        /// <returns>PoxUserInfo struct with UID and GID</returns>
+        /// <exception cref="Exception"></exception>
+        public static PosixUserInfo GetEffectiveUser(IToolLogger logger)
```
[the remark this answers is not in the document]
Comment (jasonterando) [reply src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs:33]: Ok - changing function to return nullable PosixUserInfo and not throw Exceptions

File: src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        public int UserID;
+        public int GroupID;
+        public bool UserIDSet;
+        public bool GroupIDSet;
+    }
+
+    /// <summary>
+    /// Helper class to retrieve current Linux/Mac user and group information
+    /// </summary>
+    public static class PosixUserHelper
```
[the remark this answers is not in the document]
Comment (jasonterando) [reply src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs:19]: Acknowledged

File: src/Amazon.Common.DotNetCli.Tools/DockerCLIWrapper.cs (hunk 5 above)
Comment (jasonterando) [reply src/Amazon.Common.DotNetCli.Tools/DockerCLIWrapper.cs:189]: Ok, changing this to skip if PosixUserHelper isn't able to get the user or group, or if the user and group are both zero (root):

```
                var posixUser = PosixUserHelper.GetEffectiveUser(_logger);
                if (posixUser?.UserID > 0 posixUser?.GroupID > 0)
                {
```

File: src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        /// <summary>
+        /// Return True if running under Unix or Mac
+        /// </summary>
+        public static readonly bool IsRunningInPosix = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ||
+                                                       RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
```
[the remark this answers is not in the document]
Comment (jasonterando) [reply src/Amazon.Common.DotNetCli.Tools/PosixUserHelper.cs:25]: Hi - Norm had asked me to make the inclusion of using `System.Runtime.InteropServices` contingent upon NETCOREAPP3_1_OR_GREATER being set, so that this code won't bomb under Visual Studio.  This means that I can't add that as a parameter very easily (OSPlatform is not a c# enum but a "string-in-a-struct" kind of enum).

I was also thinking it wouldn't be a bad idea of separating the "Am I *NIX?" from operations like getting the ID, in case we wanted to add more functionality. 

What I am thinking I should do isto set up an injectable class-based implementation of launching a process and getting results, which would make this testable (we could mock `id` failing or otherwise not being on the system at all).  The same Process situation exists all over this code base, and this implementation could help make that code more testable as well.

**Comment by jasonterando** [comment]
Hi, I've posted an update with changes as follows:

1. Replaced single call to "id" w/ regex parsing to separate calls to "id" for user ID and group ID
2. Failure to call "id" will be logged but ignored
3. If the user is root user and group user no changes will be made to the Docker command
4. Launching the "id" process has been re-implemented using a class to facilitate unit testing (may be reused for other launched processes); added unit tests to simulate "id" success and failure regardless of OS running unit tests
5. References to System.Runtime.InteropServices "using" statements and OS checks are wrapped with "#if NETCOREAPP3_1_OR_GREATER"

Unit and integration test all run ok.  I do not have Visual Studio to confirm things are running OK there.

**Comment by Beau-Gosse-dev** [comment]
FWIW, I tested the latest commit [625bd4f](https://github.com/aws/aws-extensions-for-dotnet-cli/pull/257/commits/625bd4f94020d4ea651f8a994c81070ee0947c4a) on Ubuntu as root and non-root and both worked. When running as root, the -u argument isn't passed, as expected.

**Review by normj — APPROVED** [review]

**Comment by normj** [comment]
@jasonterando Nothing more you need to do right now. The change looks good to me and I have run the code through its paces. I need to get somebody else on the team to double check the change and assuming nothing else found we can get the change released.

**Review by peterrsongg — APPROVED** [review]
I've tested these changes on a Linux and MacOs instance and it ran as non-root user successfully

**Comment by normj** [comment]
This was released this morning in version 5.6.3 of Amazon.Lambda.Tools
